### PR TITLE
CVE-2025-59529: simple protocol connection limits with review fixes and tests

### PR DESCRIPTION
**Pull request by abhinavagarwal07**
Picking up where #808 left off. First commit is @pemensik's connection limit fix (squashed from #808), second addresses @evverx's review and adds smoke tests, third fixes CI issues.

Commit 1 adds per-UID and global client limits to the simple protocol, derived from `RLIMIT_NOFILE`. Commit 2 fixes the review feedback from #808: unknown UID returns `(uid_t)-1` not `0`, validates `getsockopt` length, fixes format string UB on no-credential platforms, fixes the off-by-one in the per-UID check, and cleans up style. It also adds smoke tests that set `rlimit-nofile=100` (so max_clients=75, per_uid=37) and flood the socket to verify refusals are logged.

Commit 3 fixes CI: under Valgrind, `setrlimit()` is intercepted, so the test sets `ulimit -n 100` before launching Valgrind. On systemd, a transient `LimitNOFILE=100` drop-in is needed since the unit's limit overrides the daemon config. Also restarts `avahi-dnsconfd` after each daemon kill in the test (it dies when the daemon does), and fixes an off-by-one in the per-UID fast-path guard and a `max_clients=0` edge case.

**Limitations**

- `cr_pid` may need `AC_CHECK_MEMBERS` guard for FreeBSD < 13.2
- Per-UID enforcement requires credential support; without it, only the global limit applies
- Per-UID smoke test needs `runuser` (Linux only)

### CODE REVIEW

**Comment by abhinavagarwal07** [comment]
Hope this helps unblock the v0.9 release — the lack of tests for the connection limit fix was one of the open items in #503 and #808.

File: avahi-daemon/simple-protocol.c [hunk 1]
```diff
@@ -438,6 +546,56 @@ static void client_work(AvahiWatch *watch, AVAHI_GCC_UNUSED int fd, AvahiWatchEv
...
+     * treated as non-root so they cannot bypass the per-UID limit.
+     *
+     * On platforms without credential support, all connections share one
+     * "unknown UID" bucket, which effectively limits them to max_uid_clients
+     * instead of max_clients. This is an acceptable tradeoff: returning 0
+     * (root) would silently skip per-UID enforcement entirely. */
+    if (uid != 0 && n_clients > s->max_uid_clients) {
+        /* There are enough clients to reach the limit for one UID.
+         * Check this UID does not have too many connections already. */
+        Client *c;
```
Comment (abhinavagarwal07) [inline avahi-daemon/simple-protocol.c:580]: Worth noting: on platforms without credential support (NetBSD, OmniOS), all connections share one "unknown UID" bucket since they all get uid=(uid_t)-1. This means the effective limit is max_uid_clients instead of max_clients — roughly halved. I went with this because the alternative (returning uid=0) silently skipped per-UID enforcement entirely, which felt worse. Open to changing this if you'd prefer unknown UIDs to be exempt from per-UID checks.

**Comment by abhinavagarwal07** [comment]
@evverx fixed CI failures

**Comment by evverx** [comment]
I didn't get past the part where the test is skipped under Valgrind because it's already tested in non-Valgrind jobs. Is it LLM-generated?

**Comment by abhinavagarwal07** [comment]
I did use LLM assistance for this PR.

**Comment by evverx** [comment]
LLM-generated contributions are covered in https://github.com/avahi/avahi/blob/master/.github/CONTRIBUTING.md#llms. It was added to make sure that LLM-generated code is reviewed before PRs are opened. Our golden rule is that a contribution should be worth more to the project than the time it takes to review it.

Before proceeding with this can you expand on why it was decided to launch a bunch of socat processes and why  it says that "systemd's LimitNOFILE overrides the daemon config"?
